rust/xaynet-analytics/src/controller.rs

```diff
@@ -148,3 +194,392 @@ impl AnalyticsController {
+
+    #[test]
+    fn test_get_last_time_data_sent() {
+        let test_name = "test_get_last_time_data_sent_is_none";
+        let controller = get_controller(test_name, None);
+
+        let last_time_data_sent = AnalyticsController::get_last_time_data_sent(controller.db());
+        assert!(last_time_data_sent.is_ok());
+        assert!(last_time_data_sent.unwrap().is_none());
+
+        let timestamp = DateTime::parse_from_rfc3339("2021-03-03T03:03:00+00:00")
+            .unwrap()
+            .with_timezone(&Utc);
+        let controller_data = ControllerData::new(timestamp);
+        assert!(controller_data
+            .save(controller.db(), CollectionNames::CONTROLLER_DATA)
+            .is_ok());
+
+        let last_time_data_sent = AnalyticsController::get_last_time_data_sent(controller.db());
+        assert!(last_time_data_sent.is_ok());
+        assert_eq!(last_time_data_sent.unwrap(), Some(timestamp));
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_did_send_already_in_this_period_never_sent_before() {
+        let test_name = "test_did_send_already_in_this_period_never_sent_before";
+        let controller = get_controller(test_name, Some(24));
+
+        assert_eq!(controller.did_send_already_in_this_period(), false);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_did_send_already_in_this_period_inside_24h() {
+        let test_name = "test_did_send_already_in_this_period_inside_24h";
+        let initial_controller = get_controller(test_name, Some(24));
+
+        let timestamp = Utc::now();
+        let controller_data = ControllerData::new(timestamp);
+        assert!(controller_data
+            .save(initial_controller.db(), CollectionNames::CONTROLLER_DATA)
+            .is_ok());
+
+        // init controller again, to read self.last_time_data_sent from db
+        assert!(initial_controller.dispose().is_ok());
+        let controller = get_controller(test_name, Some(24));
+        assert_eq!(controller.did_send_already_in_this_period(), true);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_did_send_already_in_this_period_outside_24h() {
+        let test_name = "test_did_send_already_in_this_period_outside_24h";
+        let initial_controller = get_controller(test_name, Some(24));
+
+        let timestamp = Utc::now() - Duration::hours(25);
+        let controller_data = ControllerData::new(timestamp);
+        assert!(controller_data
+            .save(initial_controller.db(), CollectionNames::CONTROLLER_DATA)
+            .is_ok());
+
+        // init controller again, to read self.last_time_data_sent from db
+        assert!(initial_controller.dispose().is_ok());
+        let controller = get_controller(test_name, Some(24));
+        assert_eq!(controller.did_send_already_in_this_period(), false);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_did_send_already_in_this_period_inside_12h() {
```

Comment:
with the initial buggy implementation of the function being tested, it was kinda hardcoded to 24h, so it makes sense to me to check it now works with 12h. i added 6h as well because i wanted to confirm the 'sliding' kinda window implemented via the `while` loop, so that you can start 2x 6h windows away from current period.

but yes, maybe there is a bit of redundancy

Reply:
that's fine, not complaining about too many test cases!